Blog post on dashboards (#509)
flowforge/website: src/blog/2023/03/comparinig-dashboards.md

## Methodology

To compare these dashboards, I am going to consider each of them based on the following factors:

- [How easy is it to install?](#how-easy-is-it-to-install%3F)

- [How easy is it to get your first demo dashboard running?](#how-easy-is-it-to-get-your-first-demo-dashboard-running%3F)

- [How extensive is the collection of UI elements?](#how-extensive-is-the-collection-of-ui-elements%3F)

- [How good is the support and documentation?](#how-good-is-the-support-and-documentation%3F)

- [How 'cloud native' is the dashboard?](#how-'cloud-native'-is-the-dashboard%3F)

- [How active is each project's development?](#how-active-is-each-project's-development%3F)

- [What are the future development plans?](#what-are-the-future-development-plans%3F)

> I am assuming the user is a low-code developer. They may have limited experience with coding but they are most comfortable working in visual interfaces.

Reviewer: ```suggestion
I am assuming the user is a low-code developer. They may have limited experience with coding and are most comfortable working in visual interfaces.
```

> I am also assuming that the user has been recommended to search for each of these dashboards by name. It's a little unfair to say that Dashboard is inherently better because its name is an obvious search term for 'dashboards'.

Author: The point being, if you don't know what you are searching for you will always end up using Dashboard is its name is synonymous with dashboards. I can remove it if you feel it's redundant?

So, that's the methodology, let's get on with looking at the strengths of each project.

## How easy is it to install?

### uibuilder - 1st place

> A search on [DuckDuckGo for the term 'node red ui builder'](https://duckduckgo.com/?q=node+red+ui+builder&t=h_&ia=web) returns the correct custom node. When searching for the custom node in the palette manager there is only one result, this is great as users are very likely to install what they were searching for. Once you've found the correct custom node, the installation takes just a few moments using the palette manager.

Reviewer: I am curious why you are using DuckDuckGO for search. A very large majority of people use Google for search?

Author: DDG gives the same results to all users where as Google search is very inconsistent. I feel using DDG is a more valid way to understand current search rankings but I don't mind changing it.

### FlexDash - 3rd place

When [searching on DuckDuckGo for 'node red flexdash'](https://duckduckgo.com/?q=node+red+flexdash&t=h_&ia=web) the top result is the Node-RED website for the custom node. The issue with this, and this is also a problem when searching in the palette manager, is the project 'FlexDash' is apparently not what we actually need to install. When reading the readme for the project on Github it says *'You most likely do not want to explicitly install this package, you want to install the [core widgets](https://github.com/flexdash/node-red-fd-corewidgets), which will bring in this package and more and will provide a usable whole'.*

Credit to the developers for adding in this helpful text but I suspect most users will start off by installing FlexDash then later discover that was not the correct way to proceed. It would be great if the custom node which needs to be installed was the one called 'FlexDash' in my opinion.

> This problem is compounded by their being no help file at all for 'FlexDash' showing up on the Node-RED web site. I suspect that may be a deliberate attempt to get users to reconsider if they are looking at the right custom node but it was still a confusing start for me and I suspect many users will have a similar experience.

Reviewer 2: ```suggestion
This problem is compounded by their being no help file at all for 'FlexDash' showing up on the Node-RED web site. That may well be a deliberate attempt to help users get the right custom node installed, but it was still a confusing start for me and I suspect other users will have a similar experience.
```

When setting up FlexDash, one thing that wasn't immediately obvious was that I needed to restart Node-RED before the custom node showed in the palette. This step is [covered in the docs](https://flexdash.github.io/docs/quick-start/#installing-flexdash-in-node-red) but I suspect a lot of users will get stuck working out why the palette manager says the custom node is installed but nothing new has been added to the palette.

> I believe a few improvements to the install process could make FlexDash a much more popular custom node.

Reviewer 2: May wish to add something about a resolution being actively discussed here: https://github.com/node-red/node-red/issues/569#issuecomment-1460556901

Author: Added to both places I've mentioned the install problem, thanks!

### uibuilder - 1st place

![uibuilder Github Commits](./images/uibuilder-activity.png "uibuilder Github Commits")

uibuilder has has consistent commits to the project going back several years with even greater activity since the start of 2022.

### FlexDash - 2nd place

![FlexDash Github Commits](./images/flexdash-activity.png "FlexDash Github Commits")

The commits to FlexDash have been regular since mid 2022.

### Dashboard - 3rd place

![Dashboard Github Commits](./images/dashboard-activity.png "Dashboard Github Commits")

Dashboard is now in a a maintenance only state. In the words of the project lead, *'Angular 1 (the framework used to build Dashboard) is now unsupported and it's just a matter of time before there is a serious security hole raised against it, for which there will be no fix. Of course we don't use all the features of it so we may be lucky that an exploit doesn't necessarily expose us directly but it will compromise any audits people may wish to do'*.

In practice, this means that sooner or later using Dashboard might become a significant security risk.

## What are the future development plans?

### uibuilder - joint firstplace

> During the writing of this article, uibuilder released a new version with some significant new features. In the words of the project lead when talking about version 1.6, '*It feels like uibuilder really is growing up. No more apologising for not being a direct Node-RED Dashboard replacement, uibuilder has its own path*.

Reviewer 2: ```suggestion
During the writing of this article, uibuilder released a new version with some significant new features. In the words of the project lead when talking about version 6.1.0, '*It feels like uibuilder really is growing up. No more apologising for not being a direct Node-RED Dashboard replacement, uibuilder has its own path*.
```

## Conclusions

Personally, I was a little surprised by these results. I have used Dashboard for around 3 years and always found it to be a great tool for putting together quick and informative dashboards. That being said, when attempting to objectively compare it to uibuilder and FlexDash, the other two projects often are individually better in a given category. That coupled with the halt of development for Dashboard due to AngularJS being no longer supported, it's hard to recommend Dashboard for totally new Node-RED users in 2023, especially for commercial projects.

If you already use Dashboard, in a non-commercial setting you should probably continue to do so, you might find that its development slows down to a near stall due to the underlying framework now being abandoned but for at least as of right now it's a great solution to build your Node-RED dashboards in.

FlexDash is probably the best low-code solution for building dashboards in Node-RED. If you don't get blocked by the confusing install process I believe it's the one to pick up at the time of writing due to it's ongoing support and low-code interface.

> uibuilder is currently not really a low-code solution to creating dashboards but it's quickly moving in that direction. It's got some amazing features and has a good chance of ending up as the most popular solution to build dashboards in Node-RED in the long term. That being said, as of time of writing unless you are a 'coder' you will probably struggle to build useful dashboards using it.

Reviewer 2: ```suggestion
uibuilder is currently not what I would consider a truly low-code option for creating dashboards but it is moving in that direction. It has some great features and is extremely flexible so it has a good chance of ending up as the most popular solution to build dashboards in Node-RED in the long term. That being said, as of time of writing unless you are a 'coder' you will may struggle to build a dashboard using it.
```

Open to opinion on reqording.
